### PR TITLE
Issue stratisd 2134

### LINKED ISSUE #2134: Add a script to set the Stratis keys in the initramfs
This process would involve creating another executable in `src/bin/` in the stratisd repo. This executable would use existing code from the engine in stratisd to set a key in the keyring and unlock locked devices.

Motivations
---------------
* This could be used to further develop inspection and repair tools that do not require a complete instance of stratisd to be running.
* This could also be useful to users when attempting to set up something like encryption in a semi-automated way at boot.

Steps
-------

- [x] Write an executable that uses the key infrastructure in stratisd to do key operations.
- [x] Do some investigation into getting user input from dracut-based initramfs booting.
- [x] Write a prototype of a dracut module that can query the user for passphrase input and unlock the locked crypt devices.
- [x] Test that booting from this initramfs does indeed unlock root encrypted filesystems without issue at boot.

### PR DESCRIPTION
**Pull request by jbaublitz**
Supersedes #2372 
Closes #2134

### CODE REVIEW

**Review by mulkieran — CHANGES_REQUESTED** [review]
A few requests. Should you be unifying the two systemd generators into a single executable? If not possibly something else newly added?

File: Makefile [hunk 1]
```diff
@@ -294,6 +339,8 @@ verify-dependency-bounds: set-lower-bounds
 	outdated
 	release
 	set-lower-bounds
+	build-min
```
[the remark this answers is not in the document]
Comment (jbaublitz) [reply Makefile:342]: > A few requests. Should you be unifying the two systemd generators into a single executable? If not possibly something else newly added?

Whoops! This slipped my mind. I'll do that now.

**Comment by jbaublitz** [comment]
@mulkieran Okay, this should all be resolved! I'll keep an eye on CI just to make sure it passes.

**Review by mulkieran — CHANGES_REQUESTED** [review]
Please pick up final commit https://github.com/stratis-storage/stratisd/pull/2448/commits/8c05b4c20d2574b026e55a33a7d2215d84ec5060 from https://github.com/stratis-storage/stratisd/pull/2448.

File: Makefile [hunk 2]
```diff
@@ -292,8 +337,9 @@ verify-dependency-bounds: set-lower-bounds
 	install-cfg
 	license
 	outdated
-	release
 	set-lower-bounds
+	release
```
Comment (mulkieran) [inline Makefile:341]: Pls swap w/ set-lower-bounds.

**Review by mulkieran — CHANGES_REQUESTED** [review]

**Comment by jbaublitz** [comment]
Ready for review.

File: src/bin/stratis-utils.rs [hunk 3]
```diff
@@ -166,6 +171,64 @@ fn parse_args() -> Result<(), Box<dyn Error>> {
...
+                        .required(true)
+                        .help("Directory in which to write a unit file for early priority"),
+                )
+                .arg(
+                    Arg::with_name("late_priority_dir")
+                        .required(true)
+                        .help("Directory in which to write a unit file for late priority"),
+                );
+            let matches = parser.get_matches_from(&args);
+            stratis_clevis_setup_generator::generator(
```
Comment (mulkieran) [inline src/bin/stratis-utils.rs:196]: I can't make sense of three required arguments, only one of which is used.

File: src/jsonrpc/client/utils.rs [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+    (30, "GiB"),
+    (20, "MiB"),
+    (10, "KiB"),
+    (1, "B"),
+];
+
+#[allow(clippy::cast_precision_loss)]
+pub fn to_suffix_repr(size: u128) -> String {
+    SUFFIXES.iter().fold(String::new(), |acc, (div, suffix)| {
+        let div_shifted = 1 << div;
```
Comment (mulkieran) [inline src/jsonrpc/client/utils.rs:186]: ```acc``` ends up as empty string if size is 0.

**Review by mulkieran — CHANGES_REQUESTED** [review]
Two observations...also please pick up https://github.com/stratis-storage/stratisd/pull/2448/commits/8743c564319adb3e07cb77cfccb8dc9412357a42; it is trivial but makes for an easier read.

File: src/bin/stratis-utils.rs (hunk 3 above)
Comment (jbaublitz) [reply src/bin/stratis-utils.rs:196]: This is the systemd generator interface. It is mandatory to accept three positional arguments.

File: src/jsonrpc/client/utils.rs (hunk 4 above)
Comment (jbaublitz) [reply src/jsonrpc/client/utils.rs:186]: There may also be a problem with the (1, "B") value too. I think it should be (0, "B"). I'll clean this up.

**Comment by jbaublitz** [comment]
@mulkieran Take a look at 27b2c5b9bb27852ce0eb40c56c44300aa847a7c4. I've added tests, reworked it slightly, and added a special case for `size == 0`.

**Comment by jbaublitz** [comment]
I've tested booting from Stratis as a root filesystem in my test environment and so it's back in review.

**Review by mulkieran — CHANGES_REQUESTED** [review]
I think it's time for this to go in. If you're satisfied, please close this and open a new PR for final review.

**Pull request closed without merging**